mlir/lib/Analysis/Presburger/Matrix.cpp

```diff
@@ -432,4 +432,138 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
 
 MPInt IntMatrix::normalizeRow(unsigned row) {
   return normalizeRow(row, getNumColumns());
+}
+
+MPInt IntMatrix::determinant() {
+  unsigned r = getNumRows();
+  unsigned c = getNumColumns();
+  assert(r == c);
+
+  FracMatrix m(r, c);
+  for (unsigned i = 0; i < r; i++)
+    for (unsigned j = 0; j < c; j++)
+      m.at(i, j) = Fraction(at(i, j), 1);
+  
+  Fraction det = m.determinant();
+
+  return det.getAsInteger();
+}
+
+std::optional<IntMatrix> IntMatrix::integerInverse() {
+  Fraction det = Fraction(determinant(), 1);
+  FracMatrix newMat(getNumRows(), getNumColumns());
+  for (unsigned i = 0; i < getNumRows(); i++)
+    for (unsigned j = 0; j < getNumColumns(); j++)
+      newMat(i, j) = Fraction(at(i, j), 1);
+
+  std::optional<FracMatrix> fracInverse = newMat.inverse();
+
+  if (!fracInverse)
+    return {};
+
+  IntMatrix intInverse(getNumRows(), getNumColumns());
+  for (unsigned i = 0; i < getNumRows(); i++)
+    for (unsigned j = 0; j < getNumColumns(); j++)
+      intInverse(i, j) = ((*fracInverse)(i, j) * det).getAsInteger();
+
+  return intInverse;
+}
+
+FracMatrix FracMatrix::identity(unsigned dimension) {
+  return Matrix::identity(dimension);
+}
+
+Fraction FracMatrix::determinant() {
```

Comment:
is it just me or has gaussian elimination been implemented twice here, once for det and once for inverse?

Reply:
I think the best thing for now is to just make one function that returns both (the inverse can be a pointer that will be updated if it is passed, i.e., is non-null)